src/txmempool.cpp

```diff
@@ -99,96 +139,338 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
     LOCK(cs);
     mapTx[hash] = entry;
     const CTransaction& tx = mapTx[hash].GetTx();
-        for (unsigned int i = 0; i < tx.vin.size(); i++)
-            mapNextTx[tx.vin[i].prevout] = CInPoint(&tx, i);
-    BOOST_FOREACH(const JSDescription &joinsplit, tx.vjoinsplit) {
+
+    for (unsigned int i = 0; i < tx.GetVin().size(); i++)
+        mapNextTx[tx.GetVin()[i].prevout] = CInPoint(&tx, i);
+
+    BOOST_FOREACH(const JSDescription &joinsplit, tx.GetVjoinsplit()) {
         BOOST_FOREACH(const uint256 &nf, joinsplit.nullifiers) {
             mapNullifiers[nf] = &tx;
         }
     }
+
+    for(const auto& sc: tx.GetVscCcOut()) {
+        if (mapSidechains.count(sc.scId))
+            assert(mapSidechains[sc.scId].scCreationTxHash.IsNull());
+        else
+            LogPrint("cert", "%s():%d - adding [%s] in mapSidechain\n", __func__, __LINE__, sc.scId.ToString() );
+
+        mapSidechains[sc.scId].scCreationTxHash = hash;
+    }
+
+    for(const auto& fwd: tx.GetVftCcOut()) {
+        if (mapSidechains.count(fwd.scId) == 0)
+            LogPrint("cert", "%s():%d - adding [%s] in mapSidechain\n", __func__, __LINE__, fwd.scId.ToString() );
+        mapSidechains[fwd.scId].fwdTransfersSet.insert(hash);
+    }
+
     nTransactionsUpdated++;
     totalTxSize += entry.GetTxSize();
     cachedInnerUsage += entry.DynamicMemoryUsage();
     minerPolicyEstimator->processTransaction(entry, fCurrentEstimate);
+    LogPrint("sc", "%s():%d - tx [%s] added in mempool\n", __func__, __LINE__, hash.ToString() );
 
     return true;
 }
 
+bool CTxMemPool::addUnchecked(const uint256& hash, const CCertificateMemPoolEntry &entry, bool fCurrentEstimate)
+{
+    LOCK(cs);
+    mapCertificate[hash] = entry;
+    const CScCertificate& cert = mapCertificate[hash].GetCertificate();
+
+    for (unsigned int i = 0; i < cert.GetVin().size(); i++)
+        mapNextTx[cert.GetVin()[i].prevout] = CInPoint(&cert, i);
 
-void CTxMemPool::remove(const CTransaction &origTx, std::list<CTransaction>& removed, bool fRecursive)
+    if (mapSidechains.count(cert.GetScId()) )
+        assert(mapSidechains[cert.GetScId()].backwardCertificate.IsNull());
+    else
+        LogPrint("cert", "%s():%d - adding [%s] in mapSidechain\n", __func__, __LINE__, cert.GetScId().ToString() );
+
+    mapSidechains[cert.GetScId()].backwardCertificate = hash;
+           
+    nCertificatesUpdated++;
+    totalCertificateSize += entry.GetCertificateSize();
+    cachedInnerUsage += entry.DynamicMemoryUsage();
+    // TODO cert: for the time being skip the part on policy estimator, certificates currently have maximum priority
+    // minerPolicyEstimator->processTransaction(entry, fCurrentEstimate);
```

Comment:
I think, that even if certificates have maximum priority, they have an impact on `minerPolicyEstimator` estimations for CTransaction inclusion into block and "optimal" fee rate for Transactions.